File: samples/java/src/main/java/co/windly/limbosample/presentation/splash/SplashPresenter.java
package co.windly.limbosample.presentation.splash;

import androidx.annotation.NonNull;
import co.windly.limbo.presenter.queue.LimboQueuePresenter;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import java.util.concurrent.TimeUnit;
import javax.inject.Inject;
import org.jetbrains.annotations.NotNull;
import timber.log.Timber;

class SplashPresenter extends LimboQueuePresenter<SplashView> {

  //region Constructor

  @Inject
  public SplashPresenter() {
    // No-op.
  }

  //endregion

  //region Lifecycle

  @Override
  public void attachView(@NotNull SplashView view) {
    super.attachView(view);

    // Observe automatic continue.
    observeAutomaticContinue();
  }

  //endregion

  //region Automatic Continue

  @SuppressWarnings("FieldCanBeLocal")
  private static Long AUTO_CONTINUE_DELAY = 3_000L;

  private void observeAutomaticContinue() {
    addDisposable(
      Observable.timer(AUTO_CONTINUE_DELAY, TimeUnit.MILLISECONDS)
        .subscribeOn(Schedulers.computation())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(
          this::handleObserveAutomaticContinueSuccess,
          this::handleObserveAutomaticContinueError));
  }

  private void handleObserveAutomaticContinueSuccess(@NonNull Long delay) {

    // Log the fact.
    Timber.v("Automatic continue time passed: %d", delay);
    Timber.v("Navigating to main view.");

    // Navigate to main view.
    onceViewAttached(SplashView::navigateToMainView);
  }

  private void handleObserveAutomaticContinueError(@NonNull Throwable throwable) {

    // Log an error.
    Timber.e("An error occurred while processing the automatic continue delay.");
    Timber.e(throwable);

    // Navigate to main view anyway.
    onceViewAttached(SplashView::navigateToMainView);
  }

  //endregion
}

File: samples/kotlin/src/main/java/co/windly/limbosample/presentation/splash/SplashPresenter.kt
package co.windly.limbosample.presentation.splash

import co.windly.limbo.presenter.queue.LimboQueuePresenter
import co.windly.limbo.utility.reactive.observeOnUi
import co.windly.limbo.utility.reactive.subscribeOnComputation
import io.reactivex.Observable
import io.reactivex.rxkotlin.addTo
import timber.log.Timber
import java.util.concurrent.TimeUnit.MILLISECONDS
import javax.inject.Inject

class SplashPresenter @Inject constructor() : LimboQueuePresenter<SplashView>() {

  //region Lifecycle

  override fun attachView(view: SplashView) {
    super.attachView(view)

    // Observe automatic continue.
    observeAutomaticContinue()
  }

  //endregion

  //region Automatic Continue

  companion object {
    const val AUTO_CONTINUE_DELAY = 3_000L
  }

  private fun observeAutomaticContinue() {
    Observable
      .timer(AUTO_CONTINUE_DELAY, MILLISECONDS)
      .subscribeOnComputation()
      .observeOnUi()
      .subscribe(
        ::handleObserveAutomaticContinueSuccess,
        ::handleObserveAutomaticContinueError
      )
      .addTo(disposables)
  }

  private fun handleObserveAutomaticContinueSuccess(delay: Long) = onceViewAttached {

    // Log the fact.
    Timber.v("Automatic continue time passed: %d", delay)
    Timber.v("Navigating to main view.")

    // Navigate to main view.
    it.navigateToMainView()
  }

  private fun handleObserveAutomaticContinueError(throwable: Throwable) = onceViewAttached {

    // Log an error.
    Timber.e("An error occurred while processing the automatic continue delay.")
    Timber.e(throwable)

    // Navigate to main view anyway.
    it.navigateToMainView()
  }

  //endregion
}
